### ISSUE TITLE
question: Parse the response manually

### ISSUE THREAD

**Issue by rasikas** [object/Simple.OData.Client#251]
We have a problem with one of the oData call we have to work with, it returns data set as below
~~~
{
  "d": {
    "__metadata": {
      "type": "Collection(CRM_SRV.OpptPartnerFctType)"
    },
    "results": [
      {
        "__metadata": {
          "type": "CRM_SRV.OpptPartnerFctType"
        },
        "PartnerDetermineProcedure": "A",
        "PartnerFunctionCode": "00000012",
        "ChangeableFlag": true,
        "CountLow": 0,
        "CountHigh": 10,
        "PartnerFunctionName": "Sales",
        "PartnerFunctionCategory": "0005",
        "RelationshipCategory": "DEM011"
      },
      {
        "__metadata": {
          "type": "CRM_SRV.OpptPartnerFctType"
        },
        "PartnerDetermineProcedure": "A",
        "PartnerFunctionCode": "00000014",
        "ChangeableFlag": true,
        "CountLow": 1,
        "CountHigh": 1,
        "PartnerFunctionName": "Marketing",
        "PartnerFunctionCategory": "0008",
        "RelationshipCategory": "DEMH06"
      },
      ...
      ...
      
~~~

but Simple oDataclient throws an exception with following information
~~~
Microsoft.Data.OData.ODataException: A type named 'Collection(CRM_SRV.OpptPartnerFctType)' could not be resolved by the model. When a model is available, each type name must resolve to a valid type.\r\n   at Microsoft.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at Microsoft.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccess(Task task)\r\n   at Microsoft.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)\r\n   at 
~~~

I understand this is because of the element
~~~
{
    "__metadata": {
      "type": "Collection(CRM_SRV.OpptPartnerFctType)"
    },
~~~

Unfortunately this is an existing service and provider cannot change it, is there a way to intercept call before it tries to bind the result set and manually parse by ourself to avoid the exception?

Our call would be like this
~~~
var r = await ODataClient.Unbound<OpptPartnerFctType>()
  .Function("OpptPartnerFctType")
  .Set(new Dictionary<string, object> { { "TransactionType", "JKS" } })
  .ExecuteAsEnumerableAsync();
~~~

**Comment by object** [comment]
You can achieve this by implementing AfterResponse handler (in ODataClienSettings). Here's an example that removes some offending metadata elements (written for a different but similar case).

private static void AfterResponse(HttpResponseMessage responseMessage)
        {
            if (responseMessage.RequestMessage.RequestUri.LocalPath.EndsWith("$metadata"))
            {
                var metadata = responseMessage.Content.ReadAsStringAsync().Result;
                var doc = new XmlDocument();
                doc.LoadXml(metadata);

                var manager = new XmlNamespaceManager(doc.NameTable);
                manager.AddNamespace("ns", "http://schemas.microsoft.com/ado/2008/09/edm");
                var nodes = doc.SelectNodes("//ns:Annotation", manager);
                for (var i = nodes.Count - 1; i >= 0; i--)
                {
                    nodes[i].ParentNode.RemoveChild(nodes[i]);
                }
                
                nodes = doc.SelectNodes("//ns:Property[@CollectionKind]", manager);
                for (var i = 0; i < nodes.Count; i++)
                {
                    nodes[i].Attributes.Remove(nodes[i].Attributes["CollectionKind"]);
                }

                using (var stringWriter = new StringWriter())
                using (var xmlTextWriter = XmlWriter.Create(stringWriter))
                {
                    doc.WriteTo(xmlTextWriter);
                    xmlTextWriter.Flush();
                    metadata = stringWriter.GetStringBuilder().ToString();
                }

                var byteArray = Encoding.UTF8.GetBytes(metadata);
                var stream = new MemoryStream(byteArray);
                stream.Position = 0;
                var content = new StreamContent(stream);
                foreach (var header in responseMessage.Content.Headers)
                {
                    if (header.Key != "Content-Length")
                        content.Headers.Add(header.Key, header.Value);
                }
                content.Headers.Add("Content-Length", metadata.Length.ToString());
                responseMessage.Content = content;
            }
        }

**Comment by rasikas** [comment]
Thanks for the quick reply, we actually user BeforeRequest Handler for adding headers and I understand we could use similar thing with AfterResponse, Is there a way to add this handler after we create ODataClient, because this is required only for one particular call, rest of the calls during this execution is fine, so Is there a way to add this handler into ODataClientSettings by

~~~
ODataClient.ODataClientSettings.AfterResponse += handler
~~~
or should it be always when creating the ODataClientSettings object before creating ODataClient
is there a property or method we could access ODataClientSettings object from ODataClient and add this handler and remove whenever we need
thanks a lot

**Comment by object** [comment]
Currently ODataClientSettings instance is copied to the ODataClient internal session object, so it's not possible to adjust settings later. But you should be able to include in the handle code test for some runtime condition that would only invoke response interception in one particular case.

**Comment by rasikas** [comment]
Yes that's the option I have to try for now, Intention was to use this call in a $batch call with five other oData calls but due to the issue in returned result set now I have to take this out from the $batch call and issue separate call for this and parse the data manually, Your library supports $batch calls very well so I don't want to parse that response and make things more complex, so I will take this call out of the batch call and parse separately, thanks again for the reply and support
